### PR TITLE
C++: QLDoc for Constants

### PR DESCRIPTION
**Pull request by MathiasVP**
More work towards https://github.com/github/codeql-c-analysis-team/issues/40.

I've borrowed some of the information from https://www.gnu.org/software/libc/manual/html_node/Permission-Bits.html and https://www.gnu.org/software/libc/manual/html_node/Access-Modes.html.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll [hunk 1]
```diff
@@ -15,44 +20,77 @@ int parseOctal(string input) {
...
 
+/** Gets the number corresponding to the execute-or-search permission bit for other users in Unix. */
 int s_ixoth() { result = s_ixgrp().bitShiftRight(3) }
 
+/** Gets the number corresponding to the permissions `S_IROTH | S_IWOTH | S_IXOTH` in Unix. */
 int s_irwxo() { result = s_irwxg().bitShiftRight(3) }
 
+/**
+ * Gets the number that can be used in a bitwise and with the file status flag
+ * to produce a number representing the file access mode.
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll:79]:
```suggestion
 * to produce a number representing the file access mode on Linux.
```

This number and all the numbers below are Linux-specific, right? Or do they also work on Mac/BSD? The POSIX standard doesn't specify these numbers but actually [remarks that the numbers used on Linux were a mistake](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html#tag_16_357_08).

File: cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll [hunk 2]
```diff
@@ -15,44 +20,77 @@ int parseOctal(string input) {
     )
 }
 
+/** Gets the number corresponding to the set-user-ID on execute bit in Unix. */
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll:23]:
```suggestion
/** Gets the number corresponding to the "set-user-ID on execute" bit in Unix. */
```

It took me a few tries to parse this sentence. Same problem in the sentence below.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll:23]: I went back and forth quite a few times on what was most readable as well. Adding the quotes definitely improves it!

File: cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll:79]: The values of `O_RDONLY`, `O_WRONLY`, `O_RDWR` and `O_ACCMODE` on my machine and on BSD (https://github.com/freebsd/freebsd/blob/master/sys/sys/fcntl.h#L75) and linux (https://github.com/torvalds/linux/blob/master/include/uapi/asm-generic/fcntl.h) all match the numbers in the .qll file. But indeed, the values of `O_CREAT` and `O_EXCL` are Linux specific.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/commons/unix/Constants.qll:79]: I'm fine with being pragmatic here and making it only the final 2.

**Pull request merged**